pkg/proctree/proctree_feed.go

```diff
@@ -0,0 +1,257 @@
+
+	// Case 01: The child is a process (if leader == child)
+
+	if feed.ChildHash == feed.LeaderHash {
+		leader.GetExecutable().SetFeed(parent.GetExecutable().GetFeed())
+		leader.GetInterpreter().SetFeed(parent.GetInterpreter().GetFeed())
+		return nil
+	}
+
+	// Case 02: The child is a thread, and leader is the thread group leader.
+
+	thread := pt.GetOrCreateThreadByHash(feed.ChildHash)
+	thread.GetInfo().SetFeedAt(
+		TaskInfoFeed{
+			Tid:         int(feed.ChildTid),
+			Pid:         int(feed.ChildPid),
+			NsTid:       int(feed.ChildNsTid),
+			NsPid:       int(feed.ChildNsPid),
+			StartTimeNS: feed.ChildStartTime,
+			PPid:        int(feed.ParentPid),
+			NsPPid:      int(feed.ParentNsPid),
+			Uid:         0, // TODO: implement
+			Gid:         0, // TODO: implement
+		},
+		utils.NsSinceBootTimeToTime(feed.TimeStamp),
+	)
+	thread.SetParentHash(feed.ParentHash) // all threads have the same parent as the thread group leader
+	thread.SetLeaderHash(feed.LeaderHash) // thread group leader is a "process" (not a thread)
+	leader.AddThread(feed.ChildHash)      // add the thread to the thread group leade
+
+	return nil
+}
+
+type ExecFeed struct {
+	TimeStamp         uint64
+	TaskHash          uint32
+	CmdPath           string
+	PathName          string
+	Dev               uint32
+	Inode             uint64
+	Ctime             uint64
+	InodeMode         uint16
+	InterPathName     string
+	InterDev          uint32
+	InterInode        uint64
+	InterCtime        uint64
+	Interpreter       string
+	StdinType         uint16
+	StdinPath         string
+	InvokedFromKernel int32
+}
+
+// FeedFromExec feeds the process tree with an exec event.
+func (pt *ProcessTree) FeedFromExec(feed ExecFeed) error {
+	// DEBUG (TODO: remove this)
+	// file, _ := os.Open("/dev/null")
+	// file := os.Stdout
+	// fmt.Fprintf(file, "--\nExec event received:\n")
+	// fmt.Fprintf(file, "taskHash=%v\n", feed.TaskHash)
+	// fmt.Fprintf(file, "cmdPath=%v\n", feed.CmdPath)
+	// fmt.Fprintf(file, "pathName=%v\n", feed.PathName)
+	// fmt.Fprintf(file, "interPathName=%v\n", feed.InterPathName)
+	// fmt.Fprintf(file, "interpreter=%v\n", feed.Interpreter)
+	// fmt.Fprintf(file, "stdinPath=%v\n", feed.StdinPath)
+	// fmt.Fprintf(file, "dev=%v\n", feed.Dev)
+	// fmt.Fprintf(file, "interDev=%v\n", feed.InterDev)
+	// fmt.Fprintf(file, "inode=%v\n", feed.Inode)
+	// fmt.Fprintf(file, "ctime=%v\n", feed.Ctime)
+	// fmt.Fprintf(file, "interInode=%v\n", feed.InterInode)
+	// fmt.Fprintf(file, "interCtime=%v\n", feed.InterCtime)
+	// fmt.Fprintf(file, "inodeMode=%v\n", feed.InodeMode)
+	// fmt.Fprintf(file, "stdinType=%v\n", feed.StdinType)
+	// fmt.Fprintf(file, "invokedFromKernel=%v\n", feed.InvokedFromKernel)
+	// END OF DEBUG
+
+	process, procOk := pt.GetProcessByHash(feed.TaskHash)
+	_, threadOk := pt.GetThreadByHash(feed.TaskHash)
+
+	if !procOk && !threadOk {
+		logger.Debugw("process or thread not found (evicted ?)", "taskHash", feed.TaskHash)
+		return nil
```

Comment:
if it was evicted, we should repopulate the entry with the data we currently know

Reply:
https://github.com/aquasecurity/tracee/pull/3364#discussion_r1320941337

Reply:
What I mean here is that we should recreate the missing node and not return (if we return we are still missing this node)